File: examples/convergence.rs
// --8<-- [start:convergence]
fn main() {
    use goad::convergence::Convergence;
    use goad::geom::Geom;
    use goad::params::Param;
    use goad::result::GOComponent;
    use goad::settings::{self, DEFAULT_PARTICLE_REFR_INDEX};

    // Load default settings and geometry
    let settings = settings::load_default_config().unwrap();
    let geoms = Geom::load("./examples/data/hex.obj", vec![DEFAULT_PARTICLE_REFR_INDEX]).unwrap();

    // Create a convergence solver
    let mut convergence = Convergence::new(geoms, Some(settings)).unwrap();

    // Set convergence target: 1% relative SEM on asymmetry parameter
    convergence.add_target(Param::Asymmetry, 0.03);
    convergence.add_target(Param::Asymmetry, 0.02);
    convergence.add_target(Param::ScatCross, 0.02);
    convergence.add_target(Param::ExtCrossOpticalTheorem, 0.02);
    convergence.add_target(Param::DepolarizationRatio, 0.1);
    convergence.add_target(Param::LidarRatio, 0.1);

    // Optionally set max orientations as safety cap (default is 100k)
    convergence.max_orientations = 500;

    // Solve - will terminate when target is reached or max_orientations hit
    convergence.solve().unwrap();

    // Print results (using mean() and sem() methods)
    let mean = convergence.mean();
    let sem = convergence.sem();
    let asym = mean.params.asymmetry(&GOComponent::Total).unwrap_or(0.0);
    let asym_sem = sem.params.asymmetry(&GOComponent::Total).unwrap_or(0.0);
    let relative_sem = (asym_sem / asym.abs()) * 100.0;

    println!("Orientations computed: {}", convergence.count());
    println!(
        "Asymmetry: {:.4} +/- {:.4} ({:.2}% relative SEM)",
        asym, asym_sem, relative_sem
    );

    convergence.writeup();
}
// --8<-- [end:convergence]

File: docs/source/src/python/examples/convergence.py
# --8<-- [start:basic]
from goad import Settings
from goad.convergence.asymmetry import Asymmetry
from goad.convergence.base import Convergence
from goad.convergence.convergable import Tolerance

convergence = Convergence(
    Settings(geom_path="../../../examples/data/hex.obj", quiet=True),
    [Asymmetry(tolerance=Tolerance.RELATIVE, threshold=0.02)],
)
convergence.run()
# --8<-- [end:basic]

# --8<-- [start:multiple]
from goad import Settings
from goad.convergence.asymmetry import Asymmetry
from goad.convergence.base import Convergence
from goad.convergence.convergable import Tolerance
from goad.convergence.ext_cross import ExtCross

convergence = Convergence(
    Settings(
        geom_path="../../../examples/data/hex.obj",
        quiet=True,
        particle_refr_index_im=0.001,
    ),
    [
        ExtCross(tolerance=Tolerance.RELATIVE, threshold=0.02),
        Asymmetry(tolerance=Tolerance.RELATIVE, threshold=0.02),
    ],
)
convergence.run()
# --8<-- [end:multiple]

# --8<-- [start:albedo]
from goad import Settings
from goad.convergence.albedo import Albedo
from goad.convergence.base import Convergence
from goad.convergence.convergable import Tolerance

convergence = Convergence(
    Settings(
        geom_path="../../../examples/data/hex.obj",
        quiet=True,
        particle_refr_index_im=0.001,
    ),
    [Albedo(tolerance=Tolerance.RELATIVE, threshold=0.005)],
)
convergence.run()
# --8<-- [end:albedo]


# --8<-- [start:extcross]
from goad import Settings
from goad.convergence.base import Convergence
from goad.convergence.convergable import Tolerance
from goad.convergence.ext_cross import ExtCross

convergence = Convergence(
    Settings(
        geom_path="../../../examples/data/hex.obj",
        quiet=True,
        particle_refr_index_im=0.001,
    ),
    [ExtCross(tolerance=Tolerance.RELATIVE, threshold=0.02)],
)
convergence.run()
# --8<-- [end:extcross]
